logaggregator/api.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"strconv"
+
+	"github.com/flynn/flynn/logaggregator/client"
+	"github.com/flynn/flynn/pkg/ctxhelper"
+	"github.com/flynn/flynn/pkg/httphelper"
+	"github.com/flynn/flynn/pkg/syslog/rfc5424"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/julienschmidt/httprouter"
+	"github.com/flynn/flynn/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/flynn/flynn/Godeps/_workspace/src/gopkg.in/inconshreveable/log15.v2"
+)
+
+func apiHandler(agg *Aggregator) http.Handler {
+	api := aggregatorAPI{agg: agg}
+	r := httprouter.New()
+
+	r.GET("/log/:channel_id", httphelper.WrapHandler(api.GetLog))
+	return httphelper.ContextInjector(
+		"logaggregator-api",
+		httphelper.NewRequestLogger(r),
+	)
+}
+
+type aggregatorAPI struct {
+	agg *Aggregator
+}
+
+func (a *aggregatorAPI) GetLog(ctx context.Context, w http.ResponseWriter, req *http.Request) {
+	params, _ := ctxhelper.ParamsFromContext(ctx)
+	channelID := params.ByName("channel_id")
+	vals := req.URL.Query()
+
+	// TODO(bgentry): tail -f support
+
+	lines := 0
+	strLines := vals.Get("lines")
+	if strLines != "" {
+		var err error
+		lines, err = strconv.Atoi(strLines)
+		if err != nil || lines < 0 || lines > 10000 {
+			respondWithError(w, "lines", "lines must be an integer between 0 and 10000")
+			return
+		}
+	}
+
+	w.WriteHeader(200)
+	messages := a.agg.ReadLastN(channelID, lines)
+	enc := json.NewEncoder(w)
+	for _, syslogMsg := range messages {
+		if err := enc.Encode(NewMessageFromSyslog(syslogMsg)); err != nil {
```

Comment:
I definitely agree with offering SSE streams for browsers and supporting clients, though I question whether that's the right default in an API that you might want to `curl`.